File: examples/dhcp-test.py
#!/usr/bin/python3
# -*- coding: utf-8 tab-width: 4 indent-tabs-mode: t -*-

import sys
from gi.repository import GLib
from gi.repository import gdhcp


def sig_term(signum):
    global main_loop
    main_loop.quit()
    return True
   

def handle_error(error):
    if error == gdhcp.CLIENT_ERROR_NONE:
        print("dhcp client ok")
    elif error == gdhcp.CLIENT_ERROR_INTERFACE_UNAVAILABLE:
        print("Interface unavailable")
    elif error == gdhcp.CLIENT_ERROR_INTERFACE_IN_USE:
        print("Interface in use")
    elif error == gdhcp.CLIENT_ERROR_INTERFACE_DOWN:
        print("Interface down")
    elif error == gdhcp.CLIENT_ERROR_NOMEM:
        print("No memory")
    elif error == gdhcp.CLIENT_ERROR_INVALID_INDEX:
        print("Invalid index")
    elif error == gdhcp.CLIENT_ERROR_INVALID_OPTION:
        print("Invalid option")


def no_lease_cb(dhcp_client):
    global main_loop
    print("No lease available")
    main_loop.quit()


def lease_available_cb(dhcp_client):
    print("Lease available")

    address = gdhcp.client_get_address(dhcp_client)
    print("address %s" % (address))
    if address is None:
        return

	option_value = gdhcp.client_get_option(dhcp_client, gdhcp.SUBNET);
	for (list = option_value; list; list = list->next)
		print("sub-mask %s", (char *) list->data);

	option_value = gdhcp.client_get_option(dhcp_client, gdhcp.DNS_SERVER);
	for (list = option_value; list; list = list->next)
		print("domain-name-servers %s", (char *) list->data);

	option_value = gdhcp.client_get_option(dhcp_client, gdhcp.DOMAIN_NAME);
	for (list = option_value; list; list = list->next)
		print("domain-name %s", (char *) list->data);

	option_value = gdhcp.client_get_option(dhcp_client, gdhcp.ROUTER);
	for (list = option_value; list; list = list->next)
		print("routers %s", (char *) list->data);

	option_value = gdhcp.client_get_option(dhcp_client, gdhcp.HOST_NAME);
	for (list = option_value; list; list = list->next)
		print("hostname %s", (char *) list->data);



if __name__ == "__main__":
    if len(sys.args) < 2:
        print("Usage: dhcp-test <interface index>")
        sys.exit(0)

    index = int(sys.args[1])

    print("Create DHCP client for interface %d" % (index))

    dhcp_client = gdhcp.client_new(gdhcp.IPV4, index, &error)
    if dhcp_client is None:
        handle_error(error)
        sys.exit(0)

    gdhcp.client_set_send(dhcp_client, gdhcp.HOST_NAME, "<hostname>")
    gdhcp.client_set_request(dhcp_client, gdhcp.HOST_NAME)
    gdhcp.client_set_request(dhcp_client, gdhcp.SUBNET)
    gdhcp.client_set_request(dhcp_client, gdhcp.DNS_SERVER)
    gdhcp.client_set_request(dhcp_client, gdhcp.DOMAIN_NAME)
    gdhcp.client_set_request(dhcp_client, gdhcp.NTP_SERVER)
    gdhcp.client_set_request(dhcp_client, gdhcp.ROUTER)

    gdhcp.client_register_event(dhcp_client, gdhcp.CLIENT_EVENT_LEASE_AVAILABLE, lease_available_cb, None)
    gdhcp.client_register_event(dhcp_client, gdhcp.CLIENT_EVENT_NO_LEASE, no_lease_cb, CLIENT_EVENT_NO_LEASE)

    main_loop = GLib.MainLoop()

    print("Start DHCP operation")

    timer = GLib.Timer()

    gdhcp.client_start(dhcp_client, None)

    GLib.unix_signal_add(GLib.PRIORITY_DEFAULT, signal.SIGINT, sig_term, None)
    GLib.unix_signal_add(GLib.PRIORITY_DEFAULT, signal.SIGTERM, sig_term, None)

    main_loop.run()

    del timer
    gdhcp.client_unref(dhcp_client)

    sys.exit(0)

File: examples/dhcp-test.c
/* build with:
 *   gcc `pkg-config --cflags --libs gdhcp-1.0 gobject-2.0` dhcp-test.c -o dhcp-test
 */

#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <net/route.h>
#include <net/ethernet.h>
#include <linux/if_arp.h>

#include <gdhcp.h>

static GTimer *timer;

static GMainLoop *main_loop;

static void sig_term(int sig)
{
	g_main_loop_quit(main_loop);
}

static void print_elapsed(void)
{
	gdouble elapsed;

	elapsed = g_timer_elapsed(timer, NULL);

	printf("elapsed: %f seconds\n", elapsed);
}

static void handle_error(GError *error)
{
	printf("%s\n", error->message);
}

static void no_lease_cb(GDHCPClient *dhcp_client, gpointer user_data)
{
	print_elapsed();

	printf("No lease available\n");

	g_main_loop_quit(main_loop);
}

static void lease_available_cb(GDHCPClient *dhcp_client, gpointer user_data)
{
	GList *list, *option_value = NULL;
	char *address;

	print_elapsed();

	printf("Lease available\n");

	address = gdhcp_client_get_address(dhcp_client);
	printf("address %s\n", address);
	if (!address)
		return;

	option_value = gdhcp_client_get_option(dhcp_client, GDHCP_SUBNET);
	for (list = option_value; list; list = list->next)
		printf("sub-mask %s\n", (char *) list->data);

	option_value = gdhcp_client_get_option(dhcp_client, GDHCP_DNS_SERVER);
	for (list = option_value; list; list = list->next)
		printf("domain-name-servers %s\n", (char *) list->data);

	option_value = gdhcp_client_get_option(dhcp_client, GDHCP_DOMAIN_NAME);
	for (list = option_value; list; list = list->next)
		printf("domain-name %s\n", (char *) list->data);

	option_value = gdhcp_client_get_option(dhcp_client, GDHCP_ROUTER);
	for (list = option_value; list; list = list->next)
		printf("routers %s\n", (char *) list->data);

	option_value = gdhcp_client_get_option(dhcp_client, GDHCP_HOST_NAME);
	for (list = option_value; list; list = list->next)
		printf("hostname %s\n", (char *) list->data);
}

int main(int argc, char *argv[])
{
	struct sigaction sa;
	GError *error = NULL;
	GDHCPClient *dhcp_client;
	int index;

	if (argc < 2) {
		printf("Usage: dhcp-test <interface index>\n");
		exit(0);
	}

	index = atoi(argv[1]);

	printf("Create DHCP client for interface %d\n", index);

	dhcp_client = gdhcp_client_new(G_DHCP_IPV4, index, &error);
	if (!dhcp_client) {
		handle_error(error);
		exit(0);
	}

	gdhcp_client_set_send(dhcp_client, GDHCP_HOST_NAME, "<hostname>", NULL);

	gdhcp_client_set_request(dhcp_client, GDHCP_HOST_NAME);
	gdhcp_client_set_request(dhcp_client, GDHCP_SUBNET);
	gdhcp_client_set_request(dhcp_client, GDHCP_DNS_SERVER);
	gdhcp_client_set_request(dhcp_client, GDHCP_DOMAIN_NAME);
	gdhcp_client_set_request(dhcp_client, GDHCP_NTP_SERVER);
	gdhcp_client_set_request(dhcp_client, GDHCP_ROUTER);

	g_signal_connect(dhcp_client,
					 "lease-available",
					 G_CALLBACK (lease_available_cb),
					 NULL);

	g_signal_connect(dhcp_client,
					 "no-lease",
					 G_CALLBACK (no_lease_cb),
					 NULL);

	main_loop = g_main_loop_new(NULL, FALSE);

	printf("Start DHCP operation\n");

	timer = g_timer_new();

	gdhcp_client_start(dhcp_client, NULL);

	memset(&sa, 0, sizeof(sa));
	sa.sa_handler = sig_term;
	sigaction(SIGINT, &sa, NULL);
	sigaction(SIGTERM, &sa, NULL);

	g_main_loop_run(main_loop);

	g_timer_destroy(timer);

	g_object_unref(dhcp_client);

	g_main_loop_unref(main_loop);

	return 0;
}
